src/dune_rules/melange_rules.ml

```diff
@@ -34,46 +34,71 @@ let local_of_lib ~loc lib =
           (Lib_name.to_string lib_name)
       ]
 
-let js_includes ~loc ~sctx ~target_dir ~requires_link ~scope ~js_ext =
-  let open Resolve.Memo.O in
-  Command.Args.memo
-    (Resolve.Memo.args
-       (let* (libs : Lib.t list) = requires_link in
-        let project = Scope.project scope in
-        let deps_of_lib (lib : Lib.t) =
-          let dst_dir =
-            let lib_dir =
-              let info = local_of_lib ~loc lib |> Lib.Local.info in
-              Lib_info.src_dir info
-            in
-            lib_output_dir ~target_dir ~lib_dir
-          in
-          let open Memo.O in
-          let* source_modules =
-            Dir_contents.modules_of_lib sctx lib
-            >>| Option.value_exn >>| Modules.impl_only
-          in
-          let of_module m =
-            let output = make_js_name ~js_ext ~dst_dir m in
-            Dep.file (Path.build output)
-          in
-          Resolve.Memo.return
-            (List.map source_modules ~f:of_module |> Dep.Set.of_list)
+let virtual_dst_dir_and_modules sctx lib ~loc ~target_dir =
+  match Lib.implements lib with
+  | None -> Resolve.Memo.return None
+  | Some vlib ->
+    let open Resolve.Memo.O in
+    let* vlib = vlib in
+    let info =
+      let vlib = local_of_lib ~loc vlib in
+      Lib.Local.info vlib
+    in
+    let+ modules =
+      Dir_contents.modules_of_lib sctx vlib
+      |> Resolve.Memo.lift_memo >>| Option.value_exn
```

Comment:
I know `Option.value_exn` is used already in other places after calling `Dir_contents.modules_of_lib` but just out of curiosity:

From what I see in `modules_of_lib`, the value returned can only be `None` for `| External modules` case:

https://github.com/ocaml/dune/blob/60127dfc6ebfca1dae347e3395828bed7f802c41/src/dune_rules/dir_contents.ml#L473

In which cases could that happen? And could it happen when installing melange libs?

Reply:
It can only happen for libraries that are from findlib. But since melange doesn't support external libraries yet, it's not possible at all for us.